server/src/main/java/org/elasticsearch/rest/AbstractRestChannel.java

```diff
@@ -99,6 +98,26 @@ public XContentBuilder newBuilder(
         @Nullable XContentType responseContentType,
         boolean useFiltering
     ) throws IOException {
+        return newBuilder(
+            requestContentType,
+            responseContentType,
+            useFiltering,
+            org.elasticsearch.common.io.Streams.flushOnCloseStream(bytesOutput())
+        );
+    }
+
+    /**
+     * Creates a new {@link XContentBuilder} for a response to be sent using this channel. The builder's type can be sent as a parameter,
+     * through {@code responseContentType} or it can fallback to {@link #newBuilder(XContentType, boolean)} logic if the sent type value
+     * is {@code null}.
+     */
+    @Override
+    public XContentBuilder newBuilder(
```

Comment:
Had to add an override that would allow passing a user defined stream. This was the shortest path that I could see for making all the various tricky filtering and content type decisions we have in this method available to both chunked and standard encoding code. This should really get a cleanup eventually: The logic for how the lifecycle of these builders is managed is already rather complicated and with this override we add yet another piece of complexity by having chunked encoding manage memory in yet another way.